### PR TITLE
Async requests for real!

### LINKED ISSUE #21: Introduce Async/Task-based calls for the Operations
### Description

The current version of the SwaggerProvider only exposes synchronous network calls.  It would be excellent if we had Async or Task-based versions as well, for throughput reasons.
I don't know which representation (Async or Task) would be best, especially because the operations from this library are often consumed by C# code.
### Expected behavior

Asyncronous versions of the operations would be made available.
### Actual behavior

Only synchronous versions of operations are present.
### Known workarounds

You can manually manage Tasks/Asyncs of synchronous operations via the old-style BackgroundWorker/IAsyncResult mechanisms if you so wish.

### PR DESCRIPTION
**Pull request by baronfel**
Fixes #21 

This adds a new static parameter `AsyncInsteadOfTask` and associated plumbing to add new async http calls.

To test, basically every call in the test suite is Async-aware now.
- [x] Fix tests (duplicate parameter names)
- [x] Add in Task/Async switching mechanism

### CODE REVIEW

**Comment by baronfel** [comment]
Every test failure is from a 'duplicate entry in field table' error, which I suppose means some of the schemas have duplicate fields on typenames?

**Comment by baronfel** [comment]
example failing operation definition (note the reused `thingGroupName` field in both the (required) path and (optional) query:
```
{
   "/thing-groups/{thingGroupName}":{
      "delete":{
         "description":"Deletes a thing group.",
         "operationId":"DeleteThingGroup",
         "parameters":[
            {
               "in":"body",
               "name":"body",
               "required":true,
               "schema":{
                  "$ref":"#/definitions/DeleteThingGroupRequest"
               }
            },
            {
               "in":"path",
               "name":"thingGroupName",
               "required":true,
               "type":"string"
            },
            {
               "in":"query",
               "name":"thingGroupName",
               "type":"string"
            }
         ],
         "responses":{
            "200":{
               "description":"Success",
               "schema":{
                  "$ref":"#/definitions/DeleteThingGroupResponse"
               }
            },
            "480":{
               "description":"InvalidRequestException",
               "schema":{
                  "$ref":"#/definitions/InvalidRequestException"
               }
            },
            "481":{
               "description":"VersionConflictException",
               "schema":{
                  "$ref":"#/definitions/VersionConflictException"
               }
            },
            "482":{
               "description":"ThrottlingException",
               "schema":{
                  "$ref":"#/definitions/ThrottlingException"
               }
            },
            "483":{
               "description":"InternalFailureException",
               "schema":{
                  "$ref":"#/definitions/InternalFailureException"
               }
            }
         }
      }
   }
}
```

**Comment by baronfel** [comment]
I think we can get around those failures by changing how we generate the parameter names. Instead of a pure mapping across the swagger parameters, we could instead do a fold, accumulating a `string set, ProvidedParameter list` tuple of current parameter names and the provided parameters themselves. Then if we encounter a duplicate parameter name, we append the `parameter.In` to the name to disambiguate. This works because you can't duplicate names across locations.  The only caveat is remembering to reverse the list after the fold so that required parameters are kept before optional ones.

**Comment by baronfel** [comment]
@sergey-tihon  how do you feel about each Swagger operation exposing between 1-3 methods?

so for an Operation GetPet, we could generate `GetPet` (sync), `GetPetAsync` (async), and `GetPetTask`(task)?

could we perhaps generate all three as overloads?
`GetPet(string petId) -> Pet`
`GetPet(string petId) -> Async<Pet>`
`GetPet(string petId) -> Task<Pet>`

and let overload resolution do its thing? we could let a user specify an enum as a static param to determine how many of these get generated?

so `type PetStore = SwaggerProvider<..., Methods = Methods.Sync ||| Methods.Async ||| Methods.Task>` to generate all overloads?

**Comment by baronfel** [comment]
alright, I fixed the parameter name collisions, and now also made the task/async/sync static parameter that I mentioned above. I started with the same name, different return types strategy because it was the easiest to do.  Then, I had to pin the tests to use async. For that reason I'm considering switching the default to Async. It's 2018, we shouldn't block on Http Requests :D

The test failures from the previous commits are now just github api rate limiting, so if that continues I may mark that test as pending.

**Comment by sergey-tihon** [comment]
Thank you for your effort @baronfel! Do not you mid to rebase this PR on top of dev branch? It already contains fixes for parameter duplicates (I did clean up this branch today).

I do like the idea of being able to generate all 3 types of signatures  by user request, but all three by default. I think that we could always generate `GetPet` and `GetPetAsync(..)->Task<Pet>` and provide static parameter `AsyncInsteadOfTask=true`?

For `v1.0` I really want to remove the dependency on `FSharp.Data` `http.fs` implementation and provide a client class that uses `HttpClient` inside, that is why `Task<T>` is a good default. (https://github.com/RSuter/NSwag do similar things)

**Comment by sergey-tihon** [comment]
Enum is a nice idea, but I am but sure about `*Task` suffix, I never saw such method names. In the same time `*Async` is the good known naming pattern in dotnet.

**Comment by baronfel** [comment]
That all sounds fine to me. I'll rebase the branch and make the changes you suggest.

**Comment by baronfel** [comment]
ok, rebased the branch onto dev, added the `Async` suffix where appropriate, and changed the static parameter to toggle the return type of the generated async method

**Comment by sergey-tihon** [comment]
> C:\Users\appveyor\AppData\Local\Temp\1\tmp3181.fs (4,28)-(4,120) typecheck error The type provider 'SwaggerProvider.SwaggerTypeProvider' reported an error: The design-time type 'SwaggerProvider.Internal.Compilers.ReflectionHelper' utilized by a type provider was not found in the target reference assembly set '[tgt assembly SwaggerProvider.Runtime, Version=0.10.0.0, Culture=neutral;
 tgt assembly SwaggerProvider, Version=0.10.0.0, Culture=neutral;
 tgt assembly mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;
 tgt assembly FSharp.Core, Version=4.4.3.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a;
 tgt assembly System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089;
 tmp3D59]'. You may be referencing a profile which contains fewer types than those needed by the type provider you are using.

The error means that we emit quotations that use type or methods that not available in runtime, so TP SDK cannot do type rewrite for us. So `ReflectionHelper` and other methods like `cast` that is used in quotations should be in `runtime` assembly

**Comment by baronfel** [comment]
That makes perfect sense. One of the things I lost in the rebase was the introduction of that split, I believe. I'll move those functions over, fixup the callsites and check out the test results.

**Comment by baronfel** [comment]
Looks like we got there! The appveyor build seems to be hanging on tests for some reason, but they have passed so far.

File: src/SwaggerProvider.Runtime/Parser/Schema.fs [hunk 1]
```diff
@@ -145,6 +145,8 @@ type ParameterObject =
       /// Determines the format of the array if type array is used.
       CollectionFormat: CollectionFormat
     }
+    with
+        member x.UnambiguousName = sprintf "%s%A" x.Name x.In
```
Comment (sergey-tihon) [inline src/SwaggerProvider.Runtime/Parser/Schema.fs]: Does this method generate nice names? What do you think about
```fsharp
sprintf "%sIn%A" x.Name x.In
```

**Comment by sergey-tihon** [comment]
@baronfel Very cool! I am happy to merge it, I have just a couple of question.

File: .gitignore [hunk 2]
```diff
@@ -179,11 +179,12 @@ release.cmd
 release.sh
 localpackages/
 paket-files
+.paket/load/
```
[the remark this answers is not in the document]
Comment (baronfel) [reply .gitignore]: When you generate-load-scripts with paket they end up in this directory. I use them to quickly test packages in the repl

File: src/SwaggerProvider.DesignTime/SwaggerProviderConfig.fs [hunk 3]
```diff
@@ -12,14 +12,23 @@ module private SwaggerProviderConfig =
...
 
         let staticParams =
             [ ProvidedStaticParameter("Schema", typeof<string>)
-              ProvidedStaticParameter("Headers", typeof<string>,"")
+              ProvidedStaticParameter("Headers", typeof<string>, "")
               ProvidedStaticParameter("IgnoreOperationId", typeof<bool>, false)
               ProvidedStaticParameter("IgnoreControllerPrefix", typeof<bool>, true)
-              ProvidedStaticParameter("ProvideNullable", typeof<bool>, false)]
+              ProvidedStaticParameter("ProvideNullable", typeof<bool>, false)
+              ProvidedStaticParameter("AsyncInsteadOfTask", typeof<bool>, false)]
```
[the remark this answers is not in the document]
Comment (baronfel) [reply src/SwaggerProvider.DesignTime/SwaggerProviderConfig.fs]: I suppose something like `PreferAsyncOverTask` it even just `PreferAsync` would work as well. Names are hard, though :D

File: src/SwaggerProvider.Runtime/Parser/Schema.fs (hunk 1 above)
Comment (baronfel) [reply src/SwaggerProvider.Runtime/Parser/Schema.fs]: That's a nicer name : thumbsup:

**Comment by baronfel** [comment]
I think my only reservations remaining with this are that I think that we shouldn't even generate the sync versions of the API calls. It's very little overhead in the user to call . Result on a task or Async.RunSynchronously on an Async, and I think we should encourage non-blocking behaviors

File: src/SwaggerProvider.DesignTime/SwaggerProviderConfig.fs (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergey-tihon) [reply src/SwaggerProvider.DesignTime/SwaggerProviderConfig.fs]: I trust your choice, go with `PreferAsync` if you like it

**Comment by sergey-tihon** [comment]
Yes, I can accept it... It is 2018, we can do it!
Let's go with `Task` by default and `Async` when `PreferAsync=true`.

What about method names? Will we keep `Async` suffix or not?

**Comment by baronfel** [comment]
I'd say drop the Async suffix and let the operation names suffice. If someone has looked at the swagger schema then they'd expect to see those operation names.

**Comment by baronfel** [comment]
Brief comment: with `PreferNullable` we default to an F#-friendly type: `Option`. With `PreferAsync` we default to a C#-friendly type. Should we unify one way or the other?

**Comment by baronfel** [comment]
- dropped `Async`
- dropped synchronous method generation
- renamed static parameter
- changed `UnambiguousName` generation
- updated docs!

**Comment by sergey-tihon** [comment]
Thank you @baronfel !!!

**Comment by baronfel** [comment]
Glad to see this done, finally :) Thanks for the review and merge, Sergey!

**Pull request merged**